plugins/node/opentelemetry-instrumentation-undici/src/undici.ts

```diff
@@ -0,0 +1,460 @@
+  private onRequestHeaders({ request, socket }: RequestHeadersMessage): void {
+    const record = this._recordFromReq.get(request as UndiciRequest);
+
+    if (!record) {
+      return;
+    }
+
+    const config = this._getConfig();
+    const { span } = record;
+    const { remoteAddress, remotePort } = socket;
+    const spanAttributes: Attributes = {
+      [SemanticAttributes.NETWORK_PEER_ADDRESS]: remoteAddress,
+      [SemanticAttributes.NETWORK_PEER_PORT]: remotePort,
+    };
+
+    // After hooks have been processed (which may modify request headers)
+    // we can collect the headers based on the configuration
+    const rawHeaders = request.headers.split('\r\n');
+    const reqHeaders = new Map(
+      rawHeaders.map(h => {
+        const sepIndex = h.indexOf(':');
+        const name = h.substring(0, sepIndex).toLowerCase();
+        const val = h.substring(sepIndex + 1).trim();
+        return [name, val];
+      })
+    );
+
+    if (config.headersToSpanAttributes?.requestHeaders) {
+      config.headersToSpanAttributes.requestHeaders
+        .map(name => name.toLowerCase())
+        .filter(name => reqHeaders.has(name))
+        .forEach(name => {
+          spanAttributes[`http.request.header.${name}`] = reqHeaders.get(name);
+        });
+    }
+
+    span.setAttributes(spanAttributes);
+  }
+
+  // This is the 3rd message we get for each request and it's fired when the server
+  // headers are received, body may not be accessible yet.
+  // From the response headers we can set the status and content length
+  private onResponseHeaders({
+    request,
+    response,
+  }: ResponseHeadersMessage): void {
+    const record = this._recordFromReq.get(request);
+
+    if (!record) {
+      return;
+    }
+
+    const { span, attributes, startTime } = record;
+    // We are currently *not* capturing response headers, even though the
+    // intake API does allow it, because none of the other `setHttpContext`
+    // uses currently do
+    const spanAttributes: Attributes = {
+      [SemanticAttributes.HTTP_RESPONSE_STATUS_CODE]: response.statusCode,
+    };
+
+    // Get headers with names lowercased but values intact
+    const resHeaders = new Map<string, string>();
+    for (let idx = 0; idx < response.headers.length; idx = idx + 2) {
+      resHeaders.set(
+        response.headers[idx].toString().toLowerCase(),
+        response.headers[idx + 1].toString()
+      );
+    }
+
+    // Put response headers as attributes based on config
+    const config = this._getConfig();
+    if (config.headersToSpanAttributes?.responseHeaders) {
+      config.headersToSpanAttributes.responseHeaders
+        .map(name => name.toLowerCase())
+        .filter(name => resHeaders.has(name))
+        .forEach(name => {
+          spanAttributes[`http.response.header.${name}`] = resHeaders.get(name);
+        });
+    }
+
+    // `content-length` header is a special case
+    if (resHeaders.has('content-length')) {
```

Comment:
Performance nit: If ` config.headersToSpanAttributes.responseHeaders` is not given (the common case), then the only usage of the `resHeaders` Map allocation is for this content-length value. It might be nice to optimize-away creating the `resHeaders` in that case.   I'm nitpicking. This can/should totally come only after perf benchmarking shows it is significant at all.

Reply:
thinking about it a bit I think it can be done with no Map allocation at all. I'll push the new approach in the next commit